Image: Reimplement lightbox trigger as a minimal button in corner of image

artemiomorales opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
This PR changes the lightbox trigger so that it exists in the top right corner of the image rather than overlaying it.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Addresses https://github.com/WordPress/gutenberg/issues/55097
Having the lightbox overlay the entire image interrupts key browser functionality and is prone to failing with certain themes and layouts.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
The PR changes the styles of the button and removes the JavaScript handling that was calculating the button position dynamically.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->
1. Add an image to a post
2. Enable 'Expand on Click' in the image block settings
3. Publish and view the post
4. Hover over the image with a mouse and see that the lightbox trigger becomes visible in the top right corner. Notice that hovering over the button will cause the cursor to become a pointer. Try clicking on the button and ensure the lightbox opens.
5. Notice also that hovering over the image causes the cursor to become a magnifying glass. Click on the image and ensure the lightbox opens.

Additionally:
- Try adding a `title` to the image. Hover over the image and ensure the `title` appears in a tooltip.
- Try right clicking on an image and ensuring the image contextual menu appears as expected in Chrome, Safari, Firefox, Opera.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
1. While viewing the post, tab to the image and make sure the lightbox trigger appears on focus. 
2. Press `Enter` or `Spacebar` to ensure the lightbox opens as expected.

## Screenshot
<img width="1297" alt="Screenshot 2023-10-10 at 10 36 50 AM" src="https://github.com/WordPress/gutenberg/assets/5360536/983cd177-2595-40ba-b3be-2fdab688f8f0">

artemiomorales committed:
Reimplement lightbox trigger as a minimal button in corner of image

artemiomorales committed:
Remove obsolete directives

artemiomorales committed:
Update directives to fire logic properly via image or button click

artemiomorales committed:
Ensure lightbox button only appears when hovering over image, not whole figure

artemiomorales committed:
Ensure close button does not receive focus when opening lightbox via mouse

artemiomorales committed:
Ensure button only receives focus when lightbox is closed via keyboard

artemiomorales committed:
Add comments

artemiomorales committed:
Prevent unnecessary focus being shown on mobile

artemiomorales commented:
I'm noting here that if we proceed with this, as part of it, we'll also be able to add an additional style declaration that would resolve the bug wherein theme styles conflict with the lightbox button in https://github.com/WordPress/gutenberg/issues/54682

michalczaplinski commented:
I think that showing a button on hover only is a decent compromise, nice work @artemiomorales 👍 

I see 2 issues here though: 

1. When using some custom aspect ratios, the button can appear outside the image: 

   https://github.com/WordPress/gutenberg/assets/5417266/08153050-d334-4a81-83a5-531db4a14e81

2. Images with rounded borders look a bit weird having the button in the corner which is actually outside of the image. 

   <img width="500" alt="Screenshot 2023-10-11 at 16 32 53" src="https://github.com/WordPress/gutenberg/assets/5417266/66f8f338-23a7-45ae-a7df-d949627eb39f">
   
   Edge cases like this make me question a bit whether having this button is worth the trouble, but I don't want to reopen the discussion that was already had in https://github.com/WordPress/gutenberg/issues/55097. If we can fix 1., I'm fine with shipping the button in the corner of the image.

artemiomorales committed:
Add dynamic positioning for button when image uses 'contain' setting

artemiomorales commented:
@michalczaplinski Great, thanks! I've just pushed a commit covering edge cases when using `contain`. Next step is to address scenarios wherein the image has custom width or height properties.

artemiomorales committed:
WORK IN PROGRESS - Begin accounting for various edge cases

We need to account for the fact that an image can have
custom dimensions, aspect ratio, cover or contain, captions,
thumbnail dimensions, and potentially other scenarios.

This commit begins to address those issues but notably fails
in cases where one uses a horizontal image, at full scale,
with custom aspect ratio, using 'contain'.

It seems to work in all other cases that I've checked but
needs more thorough testing and the code can probably be cleaner,
and may contain some unnecessary items.

jameskoster commented on packages/block-library/src/image/style.scss:
```diff
@@ -157,25 +157,48 @@
 		position: absolute;
 		z-index: 100;
+		top: 5px;
+		right: 5px;
```
Unsure about conventions, but should this be `left` for `rtl` languages?

artemiomorales replied:
Maybe! That sounds like a good candidate for a follow up PR. I've added this point to the next version in the [lightbox tracking issue](https://github.com/WordPress/gutenberg/issues/51132).

jameskoster commented on packages/block-library/src/image/style.scss:
```diff
@@ -157,25 +157,48 @@
+		opacity: 0;
 		border: none;
-		background: none;
+		background: #1e1e1e;
```
Not sure we should use a UI color for this button. It can probably just be black to begin with. In the future we might consider making the appearance consistent with other Buttons as defined by the active theme.

richtabor replied:
Agree with black.

jameskoster commented on packages/block-library/src/image/style.scss:
```diff
@@ -157,25 +157,48 @@
 		cursor: zoom-in;
-		width: 100%;
-		height: 100%;
+		width: 25px;
```
The icon footprint should be 24px, the button can probably match.

SantosGuillamot commented on packages/block-library/src/image/view.js:
```diff
@@ -154,11 +187,27 @@ store(
 					},
+					// We need to use a handler to know whether the mouse is hovering
+					// so we know when to show the lightbox trigger button. We are unable
+					// to use just CSS for this because the button is not a child of the image.
```
Can't we use [the next-sibling combinator](https://developer.mozilla.org/en-US/docs/Web/CSS/Next-sibling_combinator) with something like this:

```CSS
img:hover + button {
    opacity: 1;
}
```

Using JS for this seems a bit complex for this use case to me.

artemiomorales replied:
You're right! I'd forgotten about that one.

SantosGuillamot commented on packages/block-library/src/image/view.js:
```diff
@@ -137,7 +135,42 @@ store(
 						);
 					},
-					hideLightbox: async ( { context } ) => {
+					// When opening the lightbox via clicking an image,
+					// we can use the event target directly and pass it to the
+					// showLightbox action, which uses it to create the styles.
+					callShowLightboxFromImage: ( {
+						context,
+						event,
+						actions,
+					} ) => {
+						actions.core.image.showLightbox(
+							{
+								context,
+								event,
+							},
+							event.target
+						);
+					},
+					// When opening the lightbox via clicking the button,
+					// we need to reach into event target's parent element to
+					// get the image element needed to create the styles.
+					callShowLightboxFromButton: ( {
+						context,
+						event,
+						actions,
+					} ) => {
+						actions.core.image.showLightbox(
+							{
+								context,
+								event,
+							},
+							// The event target we receive when clicking the button
+							// is the SVG element inside of it, so we need to go to
+							// the parent element's sibling to get the image.
+							event.target.parentElement.previousElementSibling
+						);
+					},
```
Can't we store the `<img>` element ref in the context and use it [here](https://github.com/WordPress/gutenberg/blob/cea71f1e57712d370e503e5232a806e6fda562ed/packages/block-library/src/image/view.js#L109)? Something like:

```js
setStyles( context, context.core.image.imageRef );
```

It feels like we are creating different actions serving the same purpose.

artemiomorales replied:
Yup, you're right! I've taken your suggestion and cleaned it up.

SantosGuillamot commented:
I've been taking a look at this, and I have some comments so far:

* While hovering the button cursor is `pointer` and in the image is `zoom-in`. Is this expected?

![Screenshot 2023-10-16 at 17 43 58](https://github.com/WordPress/gutenberg/assets/34552881/08ce9543-3b5b-436f-9aee-82e64145858b)

![Screenshot 2023-10-16 at 17 43 40](https://github.com/WordPress/gutenberg/assets/34552881/77b2088e-2cd6-4657-81aa-cb63469ceb71)

* In the images with the "Contain" setting activated, the cursor shows as zoom in. I believe this wasn't the case before and it was discussed back then.

![Screenshot 2023-10-16 at 17 45 37](https://github.com/WordPress/gutenberg/assets/34552881/711ac7bd-6e73-4d56-a07e-b5e05e063814)

* In small images, the button is a bit weird.

![Screenshot 2023-10-16 at 17 46 59](https://github.com/WordPress/gutenberg/assets/34552881/701b4182-c4a7-4b6e-b1fb-df7bc7b191df)

![Screenshot 2023-10-16 at 17 47 17](https://github.com/WordPress/gutenberg/assets/34552881/9462c7e7-e8c3-4d5c-88a5-6b07fbb33789)

artemiomorales commented:
@SantosGuillamot 

> While hovering the button cursor is pointer and in the image is zoom-in. Is this expected?

Yes, this is expected.

> In the images with the "Contain" setting activated, the cursor shows as zoom in. I believe this wasn't the case before and it was discussed back then.

You mean that the cursor shows as a magnifying glass even when not over the image contents, right? I'm not sure there's a good way around this. Before, we could dynamically set the button dimensions to make sure the actionable area was only the size of the contained image, but if we're not doing that, then the `img` element itself is actually just bigger than the contained image (see below — the whole highlighted area is the `img`).

A few options come to mind:
1. Don't show the magnifying glass on hover, and don't trigger the lightbox when the image is clicked.
2. Accept this as a limitation.
3. Implement a JavaScript hack that will listen for the position of the cursor and, if it's within the bounds of the image, activate an invisible `div` whose dimensions are calculated to overlay the image content precisely and which also contains the styling for the magnifying glass.

I'm leaning towards 2, and haven't decided yet whether 3 is a good idea.

<img width="513" alt="Screenshot 2023-10-16 at 2 04 00 PM" src="https://github.com/WordPress/gutenberg/assets/5360536/4d7d3cf9-d93f-4a2b-8067-158b280f8c1e">

artemiomorales committed:
Simplify and improve button placement logic

artemiomorales committed:
Simplify logic to show button on hover

artemiomorales committed:
Fix styles

artemiomorales committed:
Simplify calls to showLightbox

artemiomorales committed:
Fix style inconsistency between browsers

artemiomorales committed:
Change button position slightly

richtabor commented:
Do we need 25px offset? Cutting it to `10` reduces the display issue on smaller images: 

![CleanShot 2023-10-16 at 17 33 00](https://github.com/WordPress/gutenberg/assets/1813435/3421bbe5-c39a-423c-8882-13471b65e15e)

![CleanShot 2023-10-16 at 17 33 59](https://github.com/WordPress/gutenberg/assets/1813435/48bd8971-ff4f-4445-b617-7fa6a95adcb9)

artemiomorales commented:
@SantosGuillamot @michalczaplinski I've just pushed a few commits that now cover all of the cases that were causing issues before. I'm going to continue testing different combinations of settings but believe this is ready for another look.

The only exception is that I still haven't resolved the issue [with certain `contain` values](https://github.com/WordPress/gutenberg/pull/55212#issuecomment-1765120709). I'm not sure there's a good solution for that, or at least not one we'll be able to get it in time for the 6.4 release candidate.

richtabor commented:
Let's also use the updated fullscreen icon from #55021.

artemiomorales committed:
Reduce button offset

artemiomorales commented:
> Do we need 25px offset? Cutting it to 10 reduces the display issue on smaller images:

@richtabor Thanks for taking a look! I just cut it to 10 pixels.

artemiomorales committed:
Add style override for better consistency across themes

artemiomorales committed:
Fix logic so lightbox animates as intended; remove extraneous code

artemiomorales committed:
Update comment

SiobhyB commented:
@artemiomorales, I wanted to check whether you're still aiming to get this into today's RC1? As a reminder, the commit freeze is at 15.00 UTC today, so this would ideally need to be merged a couple of hours before that to give us time to organise the sync and commit.

artemiomorales commented:
> I wanted to check whether you're still aiming to get this into today's RC1? As a reminder, the commit freeze is at 15.00 UTC today, so this would ideally need to be merged a couple of hours before that to give us time to organise the sync and commit.

@SiobhyB Yes, that's the hope! Although it needs a review and approval from someone else at this point. I've pinged a couple of folks so we'll see if they can take a look.

SantosGuillamot commented on packages/block-library/src/image/view.js:
```diff
@@ -279,10 +282,17 @@ store(
 								];
 
-							ref.querySelector( '.close-button' ).focus();
+							// We want to avoid drawing unnecessary attention to the close
+							// button for mouse and touch users. Note that even if opening
+							// the lightbox via keyboard, the event fired is of type
+							// `pointerEvent`, so we need to rely on the `event.pointerType`
+							// property, which returns an empty string for keyboard events.
+							if ( context.core.image.pointerType === '' ) {
+								ref.querySelector( '.close-button' ).focus();
+							}
```
Why don't we want to focus the close button when opening the button by clicking the image? Right now, it keeps the active element outside of the modal, which I believe is not correct, right?

artemiomorales replied:
That was a point brought up some folks from design a while back — they did not want a close button at all because they felt it would distract and be unnecessary.

I figured adding this so the focus ring doesn't become visible and draw attention to the close button would be a good move. Personally I would prefer to not have my attention immediately drawn to a close button when trying to get a better view of an image.

artemiomorales replied:
There are potential accessibility issues here, though. The context would indeed be incorrect for folks who activate the lightbox using a mouse while also using a screen reader.

Is this a use case we should handle? I'm currently searching Google for guidance around this, and will also ping @afercia @joedolson @alexstine for feedback.

SantosGuillamot replied:
The problem is that with that approach we are losing the focus and it is causing some issues. When opening an image by clicking on it:
1. I can't close it with the ESC key.
2. I can't close it navigating with the keyboard because I can't focus on the close button.

aristath reviewed (approved):
Let me prefix this by saying I understand this is not _yet_ perfect... I can see a couple of edge-cases above that would warrant some more polishing.
However, as a whole, this implementation is a lot better than the one we currently have.
With that in mind, and understanding that we're on the clock for RC1, I'll go ahead and pre-approve this PR. This will give us the confidence that we have something better than what was previously there, while giving us another week to polish the styles based on feedback we receive from RC1 testers 👍 

The code looks good, and this is definitely a welcomed improvement

SantosGuillamot commented on packages/block-library/src/image/view.js:
```diff
@@ -154,9 +152,16 @@ store(
-								preventScroll: true,
-							} );
+
+							// We want to avoid drawing attention to the button
+							// after the lightbox closes for mouse and touch users.
+							// Note that the `event.pointerType` property returns
+							// as an empty string if a keyboard fired the event.
+							if ( event.pointerType === '' ) {
+								context.core.image.lastFocusedElement.focus( {
+									preventScroll: true,
+								} );
+							}
```
Same as with the close button: if we navigate to a lightbox with the keyboard, but I close it using a click, I lose the focus, which I assume is not correct, right?

SantosGuillamot reviewed (approved):
I've looked at it, and, apart from a couple of comments I left, I believe the PR does what is indicated in the description. The only issue I can see is the one described [here](https://github.com/WordPress/gutenberg/pull/55212#issuecomment-1765120709), but I believe we can keep it as a limitation for this RC and analyze it later.

Apart from that, I am a bit afraid of making big changes like these ones in a RC, but I'll let you decide on that.

artemiomorales commented:
> Let's also use the updated fullscreen icon from https://github.com/WordPress/gutenberg/pull/55021.

@richtabor As we were pushing the RC1 deadline, I'll change this in a follow up PR.

SiobhyB commented:
I've gone ahead to cherry-pick this to the release branch in https://github.com/WordPress/gutenberg/pull/55405/commits/7bce633cfb1162a2127b561f0e1aad5e0e7fd1e0, so it'll be included in the RC. :)

afercia commented:
I see this has been merged while I was reviewing it. No wirries, it can happen. My lunch break took more than expected I guess :) 

However, there are issues that need to be fixed, not only accessibility but also the fact that:
- `event.pointerType` is buggy across browsers
- differentiating behavior depending on pointing device vs. 'other' is not ok for accessibility

afercia commented:
So this will need a follow up for the next RC. 

- The implementation based on `event.pointerType` is buggy, as Chrome and Opera regurn an empty string as expected when using the keyboard while Firefox and Safari return `undefined`. 
- Regardless, differentiating the behavior depending on pointing device vs. 'other' is not OK for accessibility. The modal dialog focus management should always have the same behaevior regardless of mouse, touch, keyboard, other.
- The assumption that users are _always_ either using a pointing device or other is incorrect. For example, I may want to open the modal dialog with a click and close it by pewssing the Escape key.

Depending on the flow used to open the dialog _and_ on the browser in use, this is what I observed when testing:

- When opening the dialog, focus is not transferred to the dialog. As such:
  - Focus stays on the page behind the dialog. Tabbing will navigate the page content behing the modal dialog.
  - The dialgo is not announced by screen readers because focus is elsewhere.
  - The dialog doesn't close on Escape.
  - When closing the dialog, focus should always return to the trigger button, instead this is not the case.
  - As such, there is a full focus loss: tabbing again will restart the tab order from the document root.
- The trigger button doesn't visually expose its accessible name. This is a broadeer issue with all icon buttons ued on the front end and should be addressed holistically.
- The trigger button is only 24 by 24 pixels. While this is compliant with the current WCAG 2.2, I would like to see the button target size a little larger.
- When opening the modal dialog, Initial focus should go to the element with role=dialog instead of the Close button.
- The clickable image is problematic. I'm not sure why we want to make a HTML element have a non-native behavior. I'm not sure that would make it more usable in the first place, as it is a non expected beheavi, especially considering the following point.  It is even more problematic on mobile.
- Still, there is no visual hint that the image can be enlarged. It looks like a normal image and only when hovered or focusd shows a visual hint. As a user, looking at a page with a Lightbox image, I wouldn't have any clue that the image can be enlarged. See screnshot below: how users are supposed to understand the image can be enlarged?

![Screenshot 2023-10-17 at 13 12 47](https://github.com/WordPress/gutenberg/assets/1682452/62373fb1-276b-417f-9a01-9892bf9e1607)
